draft-handrews-jri.md

```diff
@@ -0,0 +1,536 @@
+
+### Identification
+
+JRI offers two identification keywords:  one for primary resources, and one for secondary resources.
+
+#### `"$id"`
+
+The value of the `"$id"` property MUST be a string, and MUST be a valid IRI-reference as defined by {{!RFC3987, Section 2.2}}, and MUST NOT contain a fragment.
+
+The object containing the `"$id"` property MUST be considered to be a primary resource and to be identified by the IRI produced by resolving the IRI-reference against the current base IRI.  This IRI MUST be considered the base IRI for the newly identified primary resource, in accordance with {{!RFC3986, Section 5.1.1}}.
+
+If the object containing the `"$id"` property is not the root object of the document, the encapsulating resource MUST be considered to be an "encapsulating entity" per {{!RFC3986, Section 5.1.2}}.
+
+#### `"$anchor"`
+
+The value of the `"$anchor"` property MUST be a string, and MUST be a valid IRI fragment according to the `ifragment` ABNF production in {{!RFC3987, Section 2.2}}.
+
+### Location
+
+Location behavior is necessary when not all locations within a data format are valid reference targets, or when object properties that appear to be JRI keywords are only actually JRI keywords in certain locations.  In formats where most locations cause automatic actions (such as applying a schema to an instance), location behavior can also create areas where the automatic actions do not apply.  These reserved locations can safely hold re-usable objects separate from any single use.
+
+The passive nature of location behavior makes it particularly suitable to be defined directly by context specifications, as noted in {{context-loc}}.  However, not all uses of JRI involve a context specification.
+
+JRI reserves the `"$defs"` keyword in order to make no-automatic-action location behavior available within the `"$"`-prefixed namespace.  This keyword also enables safe interoperable bundling by standalone tools as described in {{bundling-interop}}.[^70]
+
+[^70]: It's also possible to bundle documents that do not use `"$defs"` into a document that does, and can therefore be un-bundled by a generic bundling tool.  I'm not sure of the best way to talk about that, as I did not want to get into defining a bundling media type.
+
+#### `"$defs"`
+
+The value of the `"$defs"` property MUST be an object, which MUST have objects as the values of all of its properties.  JRI keywords MUST be allowed in these objects.  Aside from their use in JSON Pointers or similar structural identifiers, the property names within the `"$defs"` object MUST NOT be considered to impose specific semantics to the property values.
```

Comment:
Good catch!  I'll have to think on this but agree this should be fixed before merging.  Boolean schemas don't really translate.

There is some language (I think, unless I took it out) about the value of `$defs` being the same as the root object, which _would_ encompass boolean schemas if we made it something like the "root entity" (which we'd have to do if we're allowing arrays anyway... except maybe the root array isn't really the thing we'd want? meh, I will think about it).

There is something here about the values of the `$defs` object being referenceable things.  The trick is how to say that in a meaningful yet flexible way.  🤔

Reply:
What if you do what you've done in other places: JRI says it can be anything, but context specs can further restrict it?  That would cover a hypothetical system that wanted to use `$ref` and `$defs` to break up plain JSON but still allow JSON Schema to restrict the values to objects and booleans.

Reply:
Yeah, it might be that simple.  I'll try to write it up and see if it works.